docs/architecture/adr-009-revamp-testkit.md

```diff
@@ -0,0 +1,279 @@
+    fn into_consensus_state(self) -> Self::ConsensusState;
+}
+```
+
+### Decoupling IbcContext and Host environment
+
+Currently, `MockContext` implements the top validation and execution context of
+`ibc-rs`. It contains other host-specific data e.g. `host_chain_id`,
+`block_time` - that are irrelevant to the IBC context directly. If we think
+`MockContext` as a real blockchain context, the `MockContext` represents the top
+runtime - which contains the IBC module. So we implement the validation and
+execution context on `MockIbcStore`, instead of `MockContext`.
+
+With this, the `MockContext` contains two decoupled parts - the host and the IBC
+module.
+
+### Chain-like interface for `MockContext`
+
+With the above changes, we can refactor the `MockContext` to have
+blockchain-like interfaces.
+
+The `MockContext` should have `end_block`, `produce_block` and `begin_block` to
+mimic the real blockchain environment such as Cosmos-SDK.
+
+```rs
+impl<S, H> MockContext<S, H>
+where
+    S: ProvableStore + Debug,
+    H: TestHost,
+{
+    pub fn ibc_store_mut(&mut self) -> &mut MockIbcStore<S>;
+    pub fn host_mut(&mut self) -> &mut H;
+
+    pub fn generate_genesis_block(&mut self, genesis_time: Timestamp);
+    pub fn begin_block(&mut self);
+    pub fn end_block(&mut self);
+    pub fn produce_block(&mut self, block_time: Duration);
+}
+```
+
+### ICS23 compatible proof generation
+
+With the new ability of proof generation, we can now test the Tendermint light
+clients. But we need our proofs to be ICS23 compatible. ICS23 expects the IBC
+store root to be committed at a commitment prefix at a top store in the host
+environment.
+
+For this, we add an extra store in `MockContext` where the `MockIbcStore`
+commits its storage root at its
+[commitment prefix](https://github.com/cosmos/ibc-rs/blob/v0.51.0/ibc-testkit/src/testapp/ibc/core/core_ctx.rs#L127-L129)
+key.
+
+So the `MockContext` is finalized as:
+
+```rs
+pub struct MockGenericContext<S, H>
+where
+    S: ProvableStore + Debug,
+    H: TestHost
+{
+    pub main_store: S,
+    pub host: H,
+    pub ibc_store: MockIbcStore<S>,
+}
+```
+
+Now the `MockIbcStore` can generate proofs that contain the proofs in its store
+and commitment prefix. But it has to know the proofs of its commitment prefix of
+the previous heights.
+
+So we add an extra store in `MockIbcStore` to store the proofs from previous
+heights. This is similar to storing `HostConsensusState` of previous heights.
+
+```rs
+#[derive(Debug)]
+pub struct MockIbcStore<S>
+where
+    S: ProvableStore + Debug,
+{
+    ...
+    /// Map of host consensus states
+    pub host_consensus_states: Arc<Mutex<BTreeMap<u64, AnyConsensusState>>>,
+    /// Map of proofs of ibc commitment prefix
+    pub ibc_commiment_proofs: Arc<Mutex<BTreeMap<u64, CommitmentProof>>>,
+}
+```
+
+The storing of the IBC store root at the IBC commitment prefix happens in the
+end block. The storing of proofs and host consensus states happens in the
```

Comment:
Should this be `end_block` to denote that the storing of the store root and commitment prefix happens when `end_block` is called?

Reply:
Good point. Here is my view: `end_block` doesn't produce the block as the top-level store is not committed yet. So, the context doesn't have access to the latest block yet. Hence, the latest block metadata is stored after producing the block, which is the next `begin_block`.

If you agree, I should add this logic in the doc - as this is a valid question that you raised.

Reply:
So I think what you're saying is that the storing of the root at the commitment prefix happens in the _most recent_ block (i.e. not actually in a call to `end_block`)? If that's the case, it might be clearer to clarify that by saying "most recent block" instead of "end block", as that could be easily confused with the `end_block` function name.

Reply:
ah. I think I misread your comment before. Yes, it happens in the `end_block` procedure.

To avoid confusion, here is the order:

- procedure end_block
  - storing ibc_root in commitment_prefix in main_store
- procedure produce_block
  - commits the main_store
  - produces a block with its latest root
  - push the block in the blockchain
- procedure being_block
  - store host consensus state from the latest block
  - store commitment_prefix proof in the latest block